compiler/rustc_mir_build/src/thir/pattern/check_match.rs

```diff
@@ -366,14 +366,28 @@ fn unreachable_pattern(tcx: TyCtxt<'_>, span: Span, id: HirId, catchall: Option<
 }
 
 fn irrefutable_let_pattern(tcx: TyCtxt<'_>, span: Span, id: HirId, source: hir::MatchSource) {
-    tcx.struct_span_lint_hir(IRREFUTABLE_LET_PATTERNS, id, span, |lint| {
-        let msg = match source {
-            hir::MatchSource::IfLetDesugar { .. } => "irrefutable if-let pattern",
-            hir::MatchSource::WhileLetDesugar => "irrefutable while-let pattern",
-            hir::MatchSource::IfLetGuardDesugar => "irrefutable if-let guard",
-            _ => bug!(),
-        };
-        lint.build(msg).emit()
+    tcx.struct_span_lint_hir(IRREFUTABLE_LET_PATTERNS, id, span, |lint| match source {
+        hir::MatchSource::IfLetDesugar { .. } => {
+            let mut diag = lint.build("irrefutable if-let pattern");
+            diag.note("this pattern will always match, so the if-let is useless");
+            diag.help("consider removing the if-let and using a `let`");
+            diag.emit()
+        }
+        hir::MatchSource::WhileLetDesugar => {
+            let mut diag = lint.build("irrefutable while-let pattern");
+            diag.note("this pattern will always match, so the loop will never exit");
+            diag.help("consider using a `loop { ... }` and adding a `let` inside it");
```

Comment:
Maybe this instead?

```suggestion
            diag.help("consider instead using a `loop { ... }` and adding a `let` inside it");
```

Reply:
Or this?

```suggestion
            diag.help("consider instead using a `loop { ... }` with a `let` inside it");
```

I feel like this one might be the best.

Reply:
doesn't sound grammatically correct to me, how about 
```suggestion
            diag.help("consider using a `loop { ... }` instead and adding a `let` inside it");
```

Reply:
I prefer the second suggestion; `consider instead` is grammatical: it's slightly less common, but I think it reads more clearly in this case without breaking the sentence up. I'm really happy with any of these wordings though, so it's up to you.